tests/test_stateful_browser.py

```diff
@@ -351,5 +353,65 @@ def test_select_form_nr():
             browser.select_form(nr=3)
 
 
+def test_download_link(httpbin):
+    """Test downloading the contents of a link to file."""
+    browser = mechanicalsoup.StatefulBrowser()
+    browser.open(httpbin.url)
+    tmpfile = tempfile.NamedTemporaryFile()
+    current_url = browser.get_url()
+    current_page = browser.get_current_page()
+    browser.download_link(tmpfile.name, link='image/png')
+
+    # Check that the browser state has not changed
+    assert browser.get_url() == current_url
+    assert browser.get_current_page() == current_page
+
+    # Check that the file was downloaded
+    assert os.path.isfile(tmpfile.name)
+
+
+def test_download_link_404(httpbin):
+    """Test downloading the contents of a broken link."""
+    browser = mechanicalsoup.StatefulBrowser(raise_on_404=True)
+    browser.open_fake_page('<a href="/no-such-page-404">Link</a>',
+                           url=httpbin.url)
+    tmpdir = tempfile.mkdtemp()
+    tmpfile = tmpdir + '/nosuchfile.txt'
+    current_url = browser.get_url()
+    current_page = browser.get_current_page()
+    with pytest.raises(mechanicalsoup.LinkNotFoundError):
+        browser.download_link(tmpfile, link_text='Link')
+
+    # Check that the browser state has not changed
+    assert browser.get_url() == current_url
+    assert browser.get_current_page() == current_page
+
+    # Check that the file was not downloaded
+    assert not os.path.exists(tmpfile)
+
+
+def test_download_link_referer(httpbin):
+    """Test downloading the contents of a link to file."""
+    browser = mechanicalsoup.StatefulBrowser()
+    ref = httpbin.url + "/my-referer"
+    browser.open_fake_page('<a href="/headers">Link</a>',
+                           url=ref)
+    tmpfile = tempfile.NamedTemporaryFile()
+    current_url = browser.get_url()
+    current_page = browser.get_current_page()
+    browser.download_link(tmpfile.name, link_text='Link')
```

Comment:
It wasn't in my original implementation, but thinking about it some more, it would be nice for `download_link` to return the response. I don't think there are any downsides to doing this.

If we do, then we can simplify this test. No need to read the downloaded file; instead, we can simply check the headers in the returned response.

Reply:
I agree that returning the response is nice, but we must still read the file to check that the writing happened correctly. Actually, we don't in this version, but we will in my next push.

Reply:
BTW, the existing code was creating the file before the download, and then asserting that the file existed, which was the case even without the download. I'm adding one more test to check that downloading to a non-existing file works, and that downloading to an existing file overwrites the file (we may want to provide other options than overwriting, but I want to test it to see if we change the behavior later, on purpose or by accident).

Reply:
Great! Thanks for those improvements to the tests.